broadband adjoint (2.5)

tylerflex opened the pull request:
same as #1179 but rebased against 2.5, squashed, and polished up.

Note: works fine on version 2.4.2 but fails with version 2.5.0rc1 for some reason, looking into this.

Also note: there are two added tests (`_test_adjoint_broadband ...`) that should go in core.

Summary of changes:
* users can now add multiple frequencies to `output_monitors` and also different `output_monitors` can have different frequencies.
* internally, we maintain a sorted list of `freqs_adjoint` corresponding to these monitors, which become the `freq0`s of our adjoint sources.
* if not set by `JaxSimulation.fwidth_adjoint`, the `fwidth` of the adjoint sources is chosen to be smaller than some factor (0.1) times the minimum difference between the `freqs_adjoint`. This was chosen empirically to be small enough to have low error, but not be too small. When this is set, a warning is displayed to the user outlining that an `fwidth` was chosen automatically, which could lead to large `run_time` and some mitigation approaches for tightly packed frequencies.
* if not set by a new field `JaxSimulation.run_time_adjoint`, the `run_time` of the adjoint `JaxSimulation`, will be some factor (100) divided by the adjoint `fwidth`.  Again chosen empirically to avoid early shutoff in the adjoint sims, while not being too long.
* The adjoint E fields are left unnormalized initially, but then are retroactively normalized by the adjoint source `source_time` spectrum at `fwidths_adjoint`. This is because we need to apply a different normalization frequency-by-frequency depending on the `freq0` and `fwidth` of the adjoint sources, which may introduce amplitude or phase differences.
* All of the VJP calculations are vectorized over frequency. At the end of the calculations, we sum the results over the frequency axis (consistent with derivative rule: df(x,y)/dz = df/dx + dy/dy, forget the name of it.

Corresponding notebook changes: https://github.com/flexcompute-readthedocs/tidy3d-docs/pull/296

I think that covers it, thanks in advance for the review!

weiliangjin2021 commented on tidy3d/plugins/adjoint/components/simulation.py:
```diff
@@ -227,18 +225,25 @@ def _warn_if_colocate(cls, val):
+                    freqs.append(freq)
+
+        freqs.sort()
+        return freqs
```
probably fewer lines of code by using `numpy.unique`?

tylerflex replied:
thanks, made this change.

momchil-flex commented on tidy3d/components/source.py:
```diff
@@ -105,6 +105,8 @@ def spectrum(
         time_amps = time_amps[start_ind:stop_ind]
         times_cut = times[start_ind:stop_ind]
+        # TODO: this fails if tim_amps is all 0 (for example if amplitude=0.0)
+        # maybe if this case is detected, we automatically return `np.zeros_like(freqs)`?
```
Sounds like a good solution, just add it here so it doesn't fall through the cracks?

tylerflex replied:
ok added

momchil-flex commented on tidy3d/plugins/adjoint/web.py:
```diff
@@ -691,6 +694,9 @@ def run_local_bwd(
         verbose=verbose,
     )
+
+    sim_data_adj = sim_data_adj.normalize_adjoint_fields()
```
To fix the backend tests, I had to add this to the server-side adjoint handling too. Basically any time you add something to the internals in `run_local`, we need to make sure it is also incorporated in the server-side run, so let me know.

I've fixed the backend tests in a commit on your backend branch.

tylerflex replied:
I see, thanks.

tylerflex commented:
ok I just added the source thing, fixed up changelog, squashed, etc. I think this should be ready to go assuming tests all pass.

Could I leave it to you to merge the core PR? maybe with this updated submodule commit?

momchil-flex commented:
Backend tests pass, merging everything!
